velox/functions/prestosql/benchmarks/ArraySumBenchmark.cpp

```diff
@@ -0,0 +1,213 @@
+
+  auto valueTypeKind = inputArgs.front().type->childAt(0)->kind();
+  bool isCoercibleToDouble = valueTypeKind == TypeKind::TINYINT ||
+      valueTypeKind == TypeKind::SMALLINT ||
+      valueTypeKind == TypeKind::INTEGER || valueTypeKind == TypeKind::BIGINT ||
+      valueTypeKind == TypeKind::REAL || valueTypeKind == TypeKind::DOUBLE;
+  auto errorMessage =
+      std::string("Invalid value type: ") + mapTypeKindToName(valueTypeKind);
+  VELOX_USER_CHECK_EQ(isCoercibleToDouble, true, "{}", errorMessage);
+}
+
+// Create function.
+std::shared_ptr<exec::VectorFunction> create(
+    const std::string& /* name */,
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  validateType(inputArgs);
+  auto elementType = inputArgs.front().type->childAt(0);
+
+  switch (elementType->kind()) {
+    case TypeKind::TINYINT: {
+      return std::make_shared<ArraySumFunction<
+          TypeTraits<TypeKind::TINYINT>::NativeType,
+          int64_t>>();
+    }
+    case TypeKind::SMALLINT: {
+      return std::make_shared<ArraySumFunction<
+          TypeTraits<TypeKind::SMALLINT>::NativeType,
+          int64_t>>();
+    }
+    case TypeKind::INTEGER: {
+      return std::make_shared<ArraySumFunction<
+          TypeTraits<TypeKind::INTEGER>::NativeType,
+          int64_t>>();
+    }
+    case TypeKind::BIGINT: {
+      return std::make_shared<ArraySumFunction<
+          TypeTraits<TypeKind::BIGINT>::NativeType,
+          int64_t>>();
+    }
+    case TypeKind::REAL: {
+      return std::make_shared<
+          ArraySumFunction<TypeTraits<TypeKind::REAL>::NativeType, double>>();
+    }
+    case TypeKind::DOUBLE: {
+      return std::make_shared<
+          ArraySumFunction<TypeTraits<TypeKind::DOUBLE>::NativeType, double>>();
+    }
+    default: {
+      VELOX_FAIL("Unsupported Type")
+    }
+  }
+}
+
+// Define function signature.
+// array(T1) -> T2 where T1 must be coercible to bigint or double, and
+// T2 is bigint or double
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  static const std::map<std::string, std::string> s = {
+      {"tinyint", "bigint"},
+      {"smallint", "bigint"},
+      {"integer", "bigint"},
+      {"bigint", "bigint"},
+      {"real", "double"},
+      {"double", "double"}};
+  std::vector<std::shared_ptr<exec::FunctionSignature>> signatures;
+  signatures.reserve(s.size());
+  for (const auto& [argType, returnType] : s) {
+    signatures.emplace_back(exec::FunctionSignatureBuilder()
+                                .returnType(returnType)
+                                .argumentType(fmt::format("array({})", argType))
+                                .build());
+  }
+  return signatures;
+}
+
+// Register function.
+VELOX_DECLARE_STATEFUL_VECTOR_FUNCTION(udf_array_sum, signatures(), create);
+
+class ArraySumBenchmark : public functions::test::FunctionBenchmarkBase {
+ public:
+  ArraySumBenchmark() : FunctionBenchmarkBase() {
+    functions::prestosql::registerArrayFunctions();
```

Comment:
Are these needed?

Reply:
Yes, without them the benchmark fails with an exception that the "array_sum" function doesn't exist.